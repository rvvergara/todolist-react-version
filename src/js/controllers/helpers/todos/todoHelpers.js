import {
  getDataFromLocalStorage,
  setDataIntoLocalStorage,
} from '../common/storage';

const parseDate = (dateStr) => {
  const currentDate = new Date(dateStr);
  const monthPrefix = currentDate.getMonth() < 10 ? '0' : '';
  const dayPrefix = currentDate.getDate() < 10 ? '0' : '';
  return `${currentDate.getFullYear()}-${monthPrefix}${currentDate.getMonth() + 1}-${dayPrefix}${currentDate.getDate()}`;
};

export const createTodosArray = () => {
  if (getDataFromLocalStorage('todosArray') === null) {
    const todosArray = [];
    setDataIntoLocalStorage('todosArray', todosArray);
  }
};

export const getTodoDataFromForm = (name) => {
  const inputs = document.getElementsByClassName('todo-form')[0].elements;
  const title = inputs[0].value;
  const description = inputs[1].value;
  const dueDate = parseDate(inputs[2].value);
  const priority = document.getElementsByTagName('select')[0].value;
  const notes = inputs[4].value;
  return [
    title,
    description,
    dueDate,
    priority,
    notes,
    name,
  ];
};

export const updateProjectsArray = (project) => {
  const projectsArr = getDataFromLocalStorage('projectsArray');
  const projectIndex = projectsArr.findIndex(x => x.id === project.id);
  projectsArr[projectIndex] = project;
  setDataIntoLocalStorage('projectsArray', projectsArr);
  return projectsArr;
};

export const pushTodoToProject = (todo) => {
  const projectsArray = getDataFromLocalStorage('projectsArray');
  const todosArray = getDataFromLocalStorage('todosArray');
  const todoProjectInArray = projectsArray.find(project => project.id === todo.projectID);
  const parentProjectInArray = getDataFromLocalStorage(todoProjectInArray.name);
  const parentProjectStandAlone = getDataFromLocalStorage(parentProjectInArray.name);
  parentProjectInArray.todos.push(todo);
  todosArray.push(todo);
  parentProjectStandAlone.todos.push(todo);
  setDataIntoLocalStorage(todoProjectInArray.name, parentProjectStandAlone);
  updateProjectsArray(parentProjectInArray);
  setDataIntoLocalStorage('todosArray', todosArray);
};

export const updateTodoInProject = (todo, project, index) => {
  const {
    todos,
  } = project;
  todos[index] = todo;
  const todosArray = getDataFromLocalStorage('todosArray');
  const todosArrayIndex = todosArray.findIndex(t => t.id === todo.id);
  todosArray[todosArrayIndex] = todo;
  setDataIntoLocalStorage(project.name, project);
  setDataIntoLocalStorage('todosArray', todosArray);
  updateProjectsArray(project);
  return project;
};
