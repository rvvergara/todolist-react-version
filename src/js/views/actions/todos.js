import uuid from 'uuid';

const addTodo = (todo, projectName) => ({
  type: 'ADD_TODO',
  todo: {
    ...todo,
    projectName,
    id: uuid(),
  },
});

const updateTodo = (id, updates) => ({
  type: 'UPDATE_TODO',
  id,
  updates,
});

const deleteTodo = id => ({
  type: 'DELETE_TODO',
  id,
});

export {
  addTodo,
  updateTodo,
  deleteTodo,
};
