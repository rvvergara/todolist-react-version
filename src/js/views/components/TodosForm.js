import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class TodosForm extends React.Component {
  state = {
    title: this.props.todo ? this.props.todo.title : '',
    description: this.props.todo ? this.props.todo.description : '',
    dueDate: this.props.todo ? moment(this.props.todo.dueDate) : moment(),
    priority: this.props.todo ? this.props.todo.priority : 'Priority',
    notes: this.props.todo ? this.props.todo.notes : '',
    projectName: this.props.todo ? this.props.todo.projectName : '',
    done: this.props.todo ? this.props.todo.done : false,
    error: '',
    calendarFocused: false,
  };

  handleChange = (key, value) => {
    this.setState(() => ({
      [key]: value,
    }));
  }

  onDateChange = (dueDate) => {if(dueDate) this.setState(() => ({dueDate}))};

  onFocusChange = ({focused}) => this.setState(() => ({ calendarFocused: focused }));

  submitTodo = (e) => {
    const {
      title,
      description,
      dueDate,
      notes,
      priority,
      projectName,
      done,
    } = this.state;
    e.preventDefault();
    if(!title){
      this.setState({
        error: 'Please enter a title'
      });
    }else{
      this.setState({ error: ''});
      this.props.handleSubmit({
        title,
        description,
        dueDate: dueDate.valueOf(),
        notes,
        priority,
        projectName,
        done,
      });
    }
  }

  render() {
    return (
      <div>
        {
          this.state.error && <div>{this.state.error}</div>
        }
        <form
          onSubmit={this.submitTodo}
        >
          <input
            className="form-control"
            type="text"
            placeholder="Title"
            value={this.state.title}
            onChange={e => this.handleChange('title', e.target.value)}
          />
          <input
            className="form-control"
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={e => this.handleChange('description', e.target.value)}
          />
          <SingleDatePicker 
            date={this.state.dueDate}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
          />
          <select
            className="form-control"
            value={this.state.priority}
            onChange={e => this.handleChange('priority', e.target.value)}
          >
            <option value="Priority">Priority</option>
            <option value="Low">Low</option>
            <option value="Medium">Medium</option>
            <option value="High">High</option>
          </select>
          <input
            className="form-control"
            type="text"
            value={this.state.notes}
            placeholder="Notes"
            onChange={e => this.handleChange('notes', e.target.value)}
          />
          <button
            type="submit"
            className="d-none"
          >
            Add
          </button>
        </form>
      </div>
    );
  }
}

TodosForm.propTypes = {
  todo: PropTypes.instanceOf(Object),
  handleSubmit: PropTypes.func.isRequired,
}

export default TodosForm;
