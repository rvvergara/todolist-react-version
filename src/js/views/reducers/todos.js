export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_TODO':
      return [...state, action.todo];
    case 'UPDATE_TODO':
    {
      const todoIndex = state.findIndex(todo => todo.id === action.id);
      const updatedTodo = { ...state[todoIndex], ...action.updates };
      return [...state.slice(0, todoIndex), updatedTodo, ...state.slice(todoIndex + 1)];
    }
    case 'DELETE_TODO':
      return state.filter(todo => todo.id !== action.id);
    default:
      return state;
  }
};
