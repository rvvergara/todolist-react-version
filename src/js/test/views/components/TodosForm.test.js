import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import TodosForm from '../../../views/components/TodosForm';
import todos from '../../fixtures/todos';

describe('TodosForm', () => {
  let wrapper;
  let handleSubmit;
  let changeOnInput;
  let formSubmit;

  beforeEach(() => {
    handleSubmit = jest.fn();
    wrapper = shallow(
      <TodosForm
        handleSubmit={handleSubmit}
      />,
    );
    changeOnInput = (input, value) => wrapper.find(input).simulate('change', {
      target: { value },
    });
    formSubmit = () => wrapper.find('form').prop('onSubmit')({
      preventDefault: () => {},
    });
  });

  describe('rendering', () => {
    test('should render without todo', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('should render with todo', () => {
      const initialState = {
        ...todos[2],
        dueDate: moment(todos[2].dueDate),
        priority: todos[2].priority,
        error: '',
        calendarFocused: false,
        id: undefined,
      };
      wrapper = shallow(
        <TodosForm
          handleSubmit={handleSubmit}
          todo={todos[2]}
        />,
      );
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.state()).toEqual(initialState);
    });
  });

  describe('handling changes in title, description, priority, and notes', () => {
    test('should handle change in title input', () => {
      const value = 'Cool Title';
      changeOnInput('input[placeholder="Title"]', value);
      expect(wrapper.state('title')).toBe(value);
    });

    test('should handle change in description input', () => {
      const value = 'Cillum ut est aute ipsum tempor aliqua laborum culpa aliqua quis.';
      changeOnInput('input[placeholder="Description"]', value);
      expect(wrapper.state('description')).toBe(value);
    });

    test('should handle change in priority', () => {
      const value = 'High';
      changeOnInput('select', value);
      expect(wrapper.state('priority')).toBe(value);
    });

    test('should handle change in notes value', () => {
      const value = 'Velit anim magna esse pariatur pariatur excepteur qui veniam ex consectetur aute Lorem.';
      changeOnInput('input[placeholder="Notes"]', value);
      expect(wrapper.state('notes')).toBe(value);
    });
  });

  describe('interaction with SingleDatePicker', () => {
    test('should set date on date change', () => {
      const anotherDeadline = moment().add(10, 'days');
      wrapper.find('withStyles(SingleDatePicker)').prop('onDateChange')(anotherDeadline);
      expect(wrapper.state('dueDate')).toBe(anotherDeadline);
    });

    test('should set calendar focus on change', () => {
      wrapper.find('withStyles(SingleDatePicker)').prop('onFocusChange')({ focused: true });
      expect(wrapper.state('calendarFocused')).toBe(true);
    });
  });

  describe('form submission', () => {
    test('should handle invalid form submission', () => {
      formSubmit();
      expect(wrapper.state('error').length).toBeGreaterThan(0);
    });

    test('should handle successful form submission', () => {
      const value = 'Valid Title';
      changeOnInput('input[placeholder="Title"]', value);
      formSubmit();
      expect(wrapper.state('error').length).toBe(0);
      expect(handleSubmit).toHaveBeenCalled();
    });
  });
});
