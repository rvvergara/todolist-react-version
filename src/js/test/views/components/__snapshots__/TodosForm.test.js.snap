// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosForm rendering should render with todo 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Title"
      type="text"
      value="Third Task"
    />
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Fugiat consequat aliquip anim ad proident consectetur veniam."
    />
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"2019-05-07T05:46:19.367Z"}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <select
      className="form-control"
      onChange={[Function]}
    >
      <option
        value="Priority"
      >
        Priority
      </option>
      <option
        value="Low"
      >
        Low
      </option>
      <option
        value="Medium"
      >
        Medium
      </option>
      <option
        value="High"
      >
        High
      </option>
    </select>
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Notes"
      type="text"
      value="Commodo duis est occaecat esse quis nulla nostrud consequat proident ex proident non."
    />
    <button
      className="d-none"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;

exports[`TodosForm rendering should render without todo 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Title"
      type="text"
      value=""
    />
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"2019-05-03T05:46:19.424Z"}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <select
      className="form-control"
      onChange={[Function]}
      value="Priority"
    >
      <option
        value="Priority"
      >
        Priority
      </option>
      <option
        value="Low"
      >
        Low
      </option>
      <option
        value="Medium"
      >
        Medium
      </option>
      <option
        value="High"
      >
        High
      </option>
    </select>
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Notes"
      type="text"
      value=""
    />
    <button
      className="d-none"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
